### PR TITLE
fix: doctor agents check uses --json instead of TTY-only output

### LINKED ISSUE #1028: `doctor agents` check can never pass — `claude agents` invoked without a TTY, and the count measures lines not agents
### Plugin Version

11.1.0 (verified still present at `v11.3.0`)

### Claude Code Version

2.1.263

### What happened?

The `agents` category always reports:

```
⚠ Claude agents CLI returned no data
  Run 'claude agents' manually
```

This is a false positive, and it cannot ever pass on a current Claude Code. `scripts/lib/doctor.sh:1579`:

```bash
cli_output=$(claude agents 2>/dev/null | head -20 || echo "")
```

`claude agents` requires a TTY. With stdout captured by `$(...)` there isn't one, so Claude Code writes this to **stderr** and exits `0`:

```
'claude agents' requires an interactive terminal (stdout is not a TTY) — use 'claude agents --json' for a machine-readable listing.
```

`2>/dev/null` discards that notice, stdout is empty, `cli_output` is empty, and the check takes its warn branch. Since the doctor always captures this output, the condition is structural rather than environmental — the check warns 100% of the time.

The suggested remediation ("Run `claude agents` manually") does work interactively, which makes the warning look like a real finding a user should chase.

**Second defect on the same check:** even when `cli_output` is non-empty, the count is wrong. `doctor.sh:1582`:

```bash
cli_count=$(echo "$cli_output" | grep -c "^") || cli_count=0
```

That counts output *lines*, not agents, and reports them as `"Claude agents CLI: ${cli_count} agents registered"`. With `--json` the same code would report ~22 for 2 agents, so the flag change alone isn't sufficient.

### What did you expect?

The check passes when agents are registered and reports an accurate count.

### Steps to reproduce

```console
$ claude agents 2>/dev/null | head -20; echo "stdout empty -> rc=$?"
stdout empty -> rc=0

$ claude agents            # stderr notice, no stdout
'claude agents' requires an interactive terminal (stdout is not a TTY) — use 'claude agents --json' for a machine-readable listing.

$ claude agents --json     # works
[
  {
    "pid": 74206,
    "cwd": "~/…",
    "kind": "interactive",
    "startedAt": 1788835812192,
    "sessionId": "…",
    "name": "Octopus: /octo:auto",
    "status": "waiting",
    "waitingFor": "input needed"
  }
]

$ octopus doctor agents --verbose
  ⚠ Claude agents CLI returned no data
```

### Suggested fix

```diff
-        cli_output=$(claude agents 2>/dev/null | head -20 || echo "")
+        cli_output=$(claude agents --json 2>/dev/null || echo "")
```

```diff
-            cli_count=$(echo "$cli_output" | grep -c "^") || cli_count=0
+            cli_count=$(printf '%s' "$cli_output" | jq 'length' 2>/dev/null) || cli_count=0
```

Verified locally: `doctor agents --verbose` goes from 3 passed + 1 warning to **4 passed**, with a correct agent count. Note `jq` is already an assumed dependency elsewhere in the doctor (`hooks/budget-gate.sh` bails without it, and `install-deps.sh` checks for it) — if you'd rather not add a dependency to this path, `grep -c '"sessionId"'` gives the same count without `jq`.

Since `--json` is the documented machine-readable interface, it may be worth grepping for other non-interactive `claude agents` call sites.

### Which providers are configured?

Codex CLI, Cursor CLI (`cursor-agent`), Claude (host session)

### Operating System

macOS 26.3.1 (arm64), GNU bash 5.3.15

### PR DESCRIPTION
**Pull request by nyldn**
## Description
The `agents` category in `octopus doctor` always reported a false-positive warning ("Claude agents CLI returned no data") and, even when working, miscounted agents.

Root cause in `scripts/lib/doctor.sh:1577-1592` (`doctor_check_agents`):
- `claude agents` requires an interactive terminal. Captured via `$(...)`, there is no TTY, so Claude Code writes its "requires an interactive terminal" notice to **stderr**, which the check discards with `2>/dev/null`. Stdout is empty, so the check always takes the warn branch.
- Even when `cli_output` was non-empty, `grep -c "^"` counted output *lines*, not agents — wrong by nature (and with `--json` a nested/multi-line array would inflate the count further).

## Type of Change
- [x] Bug fix

## Fix
Switch to the documented non-interactive interface `claude agents --json`, and count agents from the JSON itself (`jq 'length'`, falling back to `grep -c '"sessionId"'` when `jq` is unavailable, matching the file's existing `command -v jq` fallback convention used elsewhere in `doctor.sh`).

## Testing
- `bash -n scripts/lib/doctor.sh` — syntax OK
- Functional check with a stub `claude` binary returning a 2-element JSON array: the fixed logic reports `count=2` via both the `jq` and `grep` fallback paths (see PR diff commit message for detail)
- `make ci-changed` — 3/4 suites pass (`test-doctor-v10.sh` 19/19, `test-suite-reachability.sh` 14/14, `test-windows-doctor-compat.sh` 5/5). The 4th, `test-agy-provider.sh`, fails on a timing-sensitive "stalled agy catalog lookup" test that is unrelated to this change — confirmed by stashing this diff and re-running on unmodified `origin/main`, where it fails identically (pre-existing flake, not introduced by this PR).

```bash
bash -n scripts/lib/doctor.sh
bash tests/unit/test-doctor-v10.sh
bash tests/unit/test-suite-reachability.sh
bash tests/unit/test-windows-doctor-compat.sh
```

## Related Issues
Closes #1028

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FTWQJXqrjJ7RsMFPovmEmx

---
_Generated by [Claude Code](https://claude.ai/code/session_01FTWQJXqrjJ7RsMFPovmEmx)_

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved agent registration diagnostics by validating command success and structured output.
  - Added reliable agent counting for environments with or without JSON query support.
  - Malformed, incomplete, or unavailable agent data now produces a warning instead of an incorrect pass.
  - Updated warning guidance to show the command needed to inspect registered agents manually.
  - Checks now provide more accurate results across supported agent data formats.

### CODE REVIEW

**Comment by nyldn** [comment]
Addressed the two real findings from the automated review in 8c282b26:

- **`grep -c` counts lines, not occurrences** (doctor.sh:1584-1585) — confirmed. On compact/minified single-line JSON the no-jq fallback would collapse to ≤1 regardless of actual agent count. Fixed with `grep -o '"sessionId"' | wc -l`, verified against a stubbed compact-JSON response.
- **Malformed/unparseable output silently reported as "pass: 0 agents"** (doctor.sh:1579) — confirmed. `|| cli_count=0` was masking real jq parse failures as a healthy zero-count pass. Switched to `if cli_count=$(...); then ... else warn ...` so a parse failure surfaces as a warning instead — verified this is still errexit-safe under this file's inherited `set -eo pipefail` (same class of issue as #1027).

Not acting on the other two:
- **jq `length` assumes a bare top-level array** (doctor.sh:1583) — the shape is already confirmed by #1028's own reproduction, which shows `claude agents --json` returning a bare array (`[{...}]`), not an object wrapper. No evidence in this repo contradicts that.
- **No dedicated capability gate for `--json`** (doctor.sh:1579) — flagged by the bot itself as unverified. There's no existing reference to `claude agents --json` anywhere else in the repo and no documented version boundary for when the flag was introduced, so I don't have a real version number to gate on. Guessing one would introduce an unverified claim of its own rather than fix anything. Left as `SUPPORTS_AGENTS_CLI` (v2.1.50+), same gate as before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/lib/doctor.sh [hunk 1]
```diff
@@ -1576,15 +1576,19 @@ doctor_check_agents() {
 
     if [[ "$SUPPORTS_AGENTS_CLI" == "true" ]]; then
         local cli_output
-        cli_output=$(claude agents 2>/dev/null | head -20 || echo "")
+        cli_output=$(claude agents --json 2>/dev/null || echo "")
```
[the remark this answers is not in the document]
Comment (nyldn) [reply scripts/lib/doctor.sh]: This review ran against c5b1c856, the first commit. Already fixed in 8c282b26: the jq path now uses `if cli_count=$(printf '%s' "$cli_output" | jq 'length' 2>/dev/null); then ... pass ... else ... warn ... fi`, so a real parse failure surfaces as `agents-cli: warn` instead of a masked `pass: 0 agents`. See PR comment for verification details.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/lib/doctor.sh [hunk 2]
```diff
@@ -1576,15 +1576,19 @@ doctor_check_agents() {
...
         if [[ -n "$cli_output" ]]; then
             local cli_count
-            cli_count=$(echo "$cli_output" | grep -c "^") || cli_count=0
+            if command -v jq >/dev/null 2>&1; then
+                cli_count=$(printf '%s' "$cli_output" | jq 'length' 2>/dev/null) || cli_count=0
+            else
+                cli_count=$(printf '%s' "$cli_output" | grep -c '"sessionId"') || cli_count=0
```
[the remark this answers is not in the document]
Comment (nyldn) [reply scripts/lib/doctor.sh]: This review ran against c5b1c856, the first commit. Already fixed in 8c282b26: the no-jq fallback now uses `grep -o '"sessionId"' | wc -l` to count occurrences instead of `grep -c` counting lines, so a compact single-line JSON array with multiple sessions is counted correctly. Verified against a stubbed 3-agent compact-JSON response (see PR comment).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/lib/doctor.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (nyldn) [reply scripts/lib/doctor.sh]: Good catch, this was a distinct gap from the parse-failure one. Fixed in 3d469434: `cli_output=$(claude agents --json 2>/dev/null) || cli_rc=$?` now captures the real exit status, and the check requires `$cli_rc -eq 0 && -n "$cli_output"` before reporting pass. Verified with a stub that writes valid JSON to stdout and exits 3 — the check now warns instead of passing, and normal success is unaffected. No errexit trip (this file is sourced under `set -eo pipefail`).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Re the multi-agent review (ran against 8c282b26, before the exit-status fix already pushed as 3d469434):

- **No-jq fallback masks parse failures the jq branch was hardened against** — confirmed, this was still a real gap after 8c282b26. Fixed in 5d4a10bc: the grep fallback is now gated on `cli_output` looking like a JSON array (`[[ "$cli_output" == \[* ]]`), so non-array/error output now warns instead of reporting a confident `pass: 0 agents registered`. Verified with a stub returning a plain error string.
- **Fallback count assumes the `sessionId` key/shape** and **no CLI version/flag gate for `--json`** — not acting on these, same reasoning as earlier in the thread: the schema (bare array, `sessionId` field) is confirmed by #1028's own reproduction output, and there's no version data anywhere in this repo to pin a real gate on without fabricating one.

Current head is 5d4a10bc.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Addressing the multi-agent review round (ran against 3d469434):

**Fixed in ce000c88 — this one is real and serious:**
- **Unguarded pipeline can abort the whole doctor run under `set -eo pipefail`** — confirmed and reproduced. `cli_count=$(... | grep -o '"sessionId"' | wc -l | tr -d ...)` was a bare assignment, not exempted from errexit like the jq branch. `grep -o` exits 1 on zero matches — a completely normal "0 agents registered" state — and under `pipefail` that silently aborted the entire `doctor` command mid-run, before later categories or the final report ran. Same class of bug as #1027. Added the `|| cli_count=0` fallback already used elsewhere in this same function. Reproduced the abort locally before the fix and confirmed it's gone after.

**Not acting on the rest:**
- **sessionId-count semantics diverge from `jq 'length'`** — same schema-assumption discussion as earlier in this thread; #1028's own repro confirms a bare array with one `sessionId` per record.
- **Bypasses `--live` gate** — explicitly flagged by the reviewer itself as pre-existing (predates this diff, the old code also called `claude agents` unconditionally). Out of scope for this bug-fix PR; a separate issue if it needs addressing.
- **No test coverage for the 3-way branch** — fair nit, but adding new test infrastructure is beyond a minimal bug-fix scope for this PR. All branches were verified manually with stubbed `claude` binaries (documented in this thread). Worth a follow-up issue if wanted.
- **Feature-flag comment says "claude agents list"** — that's in `providers.sh`, an unrelated file/line not touched by this diff, and appears to be describing behavior informally rather than a literal subcommand name.
- **"returned no data" conflates failure modes** — cosmetic message wording, not a functional defect; leaving as-is to keep this fix minimal.

Current head is ce000c88.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Re the latest multi-agent review (ran against a365f83f, Chris's more thorough validate-before-counting commit):

**Fixed in 4d3be71a:**
- **`malformed` mock mode defined but never exercised by a test** — fair, cheap gap to close. Added two tests calling `run_agents_check malformed` (plain `not-json` output): one confirming the jq path's `jq -e` parse failure correctly falls to `warn`, one confirming the no-jq path's bracket-shape check does the same. All 26 doctor tests pass.

**Not acting on the rest (same reasoning as earlier in this thread):**
- **No-jq fallback assumes an unverified `sessionId` schema** — #1028's own reproduction confirms the array-of-objects-with-`sessionId` shape.
- **No-jq fallback validates array shape by prefix/suffix, not real structure** — inherent limitation of not having a JSON parser without `jq`; the reviewer's own assessment calls this "an acceptable trade-off," and the code already documents it in a comment.

Current head is 4d3be71a. CI was green/in-progress with no failures as of the last check.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test-doctor-v10.sh [hunk 3]
```diff
@@ -187,6 +187,127 @@ else
...
+fi
+
+test_case "agents CLI no-jq fallback rejects genuinely malformed (non-JSON) output"
+command() {
+    if [[ "${1:-}" == "-v" && "${2:-}" == "jq" ]]; then
+        return 1
+    fi
+    builtin command "$@"
+}
+run_agents_check malformed
```
[the remark this answers is not in the document]
Comment (nyldn) [reply tests/unit/test-doctor-v10.sh:330]: Good catch — the original `not-json` fixture was trivially rejected by the outer-bracket check on both paths, so it never actually reached the jq parse or the sessionId-count fallback. Fixed in 5ff16546: added a `bracketed_malformed` mock mode emitting `[not-json]` and two tests confirming the jq path still correctly fails to parse it (warn) while the no-jq path's documented heuristic limit produces `pass: 0 agents registered` as you predicted. All 28 doctor tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Re the latest multi-agent review (ran against 4d3be71a, before the bracketed-invalid-JSON test landed in 5ff16546):

**Fixed in afc81d59:**
- **Empty-array case only tested through the forced no-jq fallback, never the actual jq path** — fair. Added a test running `empty` mode (`[]`) through the default jq path, directly verifying `jq -e`'s exit-status semantics hold for a `length` of `0` (jq treats `false`/`null` as failure, not `0`, but that guarantee was previously unverified here). All 29 doctor tests pass.

**Not acting on the rest (same schema/heuristic discussion as earlier in this thread):**
- **No-jq fallback's `sessionId` count is unverified/can drift** and **bracket-shaped garbage still counts to 0** — inherent to the no-`jq` heuristic, already documented in a code comment, and now has explicit test coverage (5ff16546) demonstrating exactly this documented limitation rather than hiding it.
- **Assumes a bare top-level array rather than `{"agents": [...]}`** — #1028's own reproduction output confirms the bare-array shape; no evidence anywhere contradicts it.

Current head is afc81d59.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Re the latest multi-agent review (ran against 5ff16546, before the jq-path empty-array test landed in afc81d59):

**Finding #1 ("`jq -e` treats a valid empty array as failure") is factually incorrect** — verified empirically, not just by re-reading docs:
```
$ printf '%s' '[]' | jq -e 'length'
0
$ echo $?
0
```
`jq -e` only treats `false` and `null` as failure; a numeric `0` result exits 0. This is exactly what my test added in afc81d59 (`agents CLI jq path reports 0 for an empty array (jq -e exit-status edge case)`) verifies directly, and it passes. No code change needed here — the existing implementation is correct.

**Not acting on the rest:**
- **No-jq fallback passes on bracket-shaped invalid JSON** / **`sessionId` count is a heuristic** — same accepted, documented, test-covered tradeoff discussed repeatedly in this thread. The review's own text acknowledges "this is explicitly exercised and accepted by the test... so it's a known, intentional tradeoff."
- **No indication the no-jq count is approximate** — a message-wording suggestion ("Consider noting..."), not a functional defect; keeping this fix minimal.
- **Duplicated pass/warn branches between jq and no-jq paths** — a refactor suggestion, not a bug; out of scope for a targeted bug-fix PR.
- **Unversioned `--json` flag reliance** — explicitly marked pre-existing by the reviewer itself and already discussed earlier in this thread: no version data exists anywhere in this repo to pin a real capability gate on.

No new commit for this round — current head remains afc81d59.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
This review round (against the current head, afc81d59) repeats the same three findings already discussed and declined multiple times earlier in this thread, with no new information:
- No-jq fallback passes on bracket-shaped invalid JSON — documented, intentional, test-covered tradeoff of not having a JSON parser without `jq`.
- No-jq fallback's `sessionId` field assumption — confirmed correct by #1028's own reproduction output.
- Unversioned `--json` flag — no version data exists anywhere in this repo to pin a real capability gate on without fabricating one.

No code changes for this round. Holding here unless new information surfaces — happy to revisit any of these if someone can point to the actual `claude agents --json` schema/version contract.

---
_Generated by [Claude Code](https://claude.ai/code)_
